### PR TITLE
graph superset and dictionary types

### PR DESCRIPTION
**Pull request by baywet**
- adds a superset for Graph

TODO:
- [x] dictionary type

### CODE REVIEW

File: graph/Guidelines.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+- [OData reference](https://github.com/oasis-tcs/odata-vocabularies/blob/master/vocabularies/Org.OData.Core.V1.md#dictionary)
+
+### When to use dictionary types
+
+Before using a dictionary type in your API definition you should make sure your scenario fits the following criteria:
+
+- The data set is dynamic and you don't know in advance how many entries could be part of that data set.
+- The data values are related to one another semanticaly.
+- The data values do not need to be individually addressable for CRUD operations.
```
Comment (mikepizzo) [inline graph/Guidelines.md]: Why does this restriction exist?  It should be perfectly valid to address an individual element from the dictionary -- i.e., myEntity/AdditionalProperties/EyeColor

Comment (baywet) [reply graph/Guidelines.md]: Shouldn't you use an entity set then if you want things to be individually addressable?

File: graph/Guidelines.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+| Date        | Version | Notes                                               |
+| ----------- | ------- | --------------------------------------------------- |
+| 2021-Mar-31 | v3.3    | 1st public release of the Graph REST API Guidelines |
+
+## Introduction
+
+The Microsoft Graph REST API guidelines are an extension of the [Microsoft REST API guidelines][1]. Readers of this document are assumed to be also reading the [Microsoft REST API guidelines][1] and be familiar with them. Microsoft Graph guidance is a superset of the Microsoft API guidelines and services should follow them *except* where this document outlines specific differences or exceptions to those guidelines. This document does contain additional Microsoft Graph specific guidance and additional details.
+
+## Dictionary types
```
Comment (baywet) [inline graph/Guidelines.md]: suggestion from Dan, add a table of content

Comment (baywet) [reply graph/Guidelines.md]: we should probably duplicate the table of content from the main one and remove the ones that don't apply

File: graph/Guidelines.md (hunk 1 above)
Comment (mikepizzo) [reply graph/Guidelines.md]: If I have a schematized collection then yes, I would use an entityset/collection-valued navigation property.

---
In reply to: [601332605](https://github.com/microsoft/api-guidelines/pull/226#discussion_r601332605) [](ancestors = 601332605)

File: graph/Guidelines.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+Dictionaries, sometimes called maps, are a collection of name-value pairs. They allow dynamic data sets to be accessed in a systematic maner and are a good compromise between a stricly defined ahead of time structure with all its named properties and between a loosely defined dynamic object (i.e. OData OpenTypes).
```
Comment (baywet) [inline graph/Guidelines.md]: this whole subject needs to it's own folder and the main files needs to be an index.

Comment (baywet) [inline graph/Guidelines.md]:
```suggestion
Dictionaries, sometimes called maps, are a collection of name-value pairs. They allow dynamic data sets to be accessed in a systematic manner and are a good compromise between a strictly defined ahead of time structure with all its named properties and between a loosely defined dynamic object (i.e. OData OpenTypes).
```

File: graph/Guidelines.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+More information:
+
+- [OData reference](https://github.com/oasis-tcs/odata-vocabularies/blob/master/vocabularies/Org.OData.Core.V1.md#dictionary)
+
+### When to use dictionary types
+
+Before using a dictionary type in your API definition you should make sure your scenario fits the following criteria:
+
+- The data set is dynamic and you don't know in advance how many entries could be part of that data set.
+- The data values are related to one another semanticaly.
```
Comment (baywet) [inline graph/Guidelines.md]:
```suggestion
- The data values are related to one another semantically. i.e.: they must derive from a common type that is not Entity.
```

File: graph/Guidelines.md (hunk 1 above)
Comment (baywet) [reply graph/Guidelines.md]: addressed on the call, it is possible and the sentence has been removed. resolving.

File: graph/dictionary/index.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+{
+  "domain": "contoso"
+}
+```
+
+### Updating the dictionary
+
+```HTTP
+PATCH https://graph.microsoft.com/v1.0/users/10/roles
```
Comment (baywet) [inline graph/dictionary/index.md:120]: @mikepizzo can you confirm I'm using the right verb and I'm not supposed to document PUT here please?

File: graph/dictionary/index.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+<ComplexType Name="assignedRoleGroupDictionary" BaseType="Org.OData.Core.V1.Dictionary">
+  <!-- Note: Strongly-typed dictionary
+  of roleSettings
+  keyed by name of roleGroup. -->
+  <Annotation Term="Org.OData.Validation.V1.OpenPropertyTypeConstraint">
+    <Collection>
+      <String>microsoft.graph.roleSettings</String>
+    </Collection>
+  </Annotation>
+  <Annotation Term="SupportedHttpMethod">
```
Comment (baywet) [inline graph/dictionary/index.md:174]: @mikepizzo calling your attention on this one

**Review by darrelmiller — APPROVED** [review]

**Pull request merged**